### ISSUE TITLE
Node names without "tower" prefix

### ISSUE THREAD

**Issue by zhou-git** [tensorpack/tensorpack#1153]
Hi,

I am trying to debug a customized model built using examples/FasterRCNN as template. Thanks to your help #1142 , now I can run the debugger.
Since my implementation may be incorrect, I also run the debugger using original FasterRCNN. Looking at the names of the nodes in the graph, it is a little confusing to me that, for the same layer, not all nodes have the prefix "tower" before them. For example, for the batchnorm layer in backbone/group1/block1/conv1, there are 29 nodes in toal(as far as I can tell) :
```
(9445 / 12861)  [      ] tower0/gradients/tower0/group1/block1/conv1/bn/batchnorm/mul_1_grad/Shape                                                             
(9446 / 12861)  [      ] tower0/gradients/tower0/group1/block1/conv1/bn/batchnorm/mul_1_grad/BroadcastGradientArgs                                             
(9447 / 12861)  [      ] group1/block1/conv1/bn/beta                                                                                                           
(9448 / 12861)  [      ] group1/block1/conv1/bn/beta/read                                                                                                      
(9449 / 12861)  [      ] tower0/group1/block1/conv1/bn/Reshape_1/shape                                                                                         
(9450 / 12861)  [      ] tower0/group1/block1/conv1/bn/
(9451 / 12861)  [      ] group1/block1/conv1/bn/mean/EMA                                                                                                       
(9452 / 12861)  [      ] group1/block1/conv1/bn/mean/EMA/read                                                                                                  
(9453 / 12861)  [      ] tower0/group1/block1/conv1/bn/Reshape_2/shape                                                                                         
(9454 / 12861)  [      ] tower0/group1/block1/conv1/bn/Reshape_2                                                                                               
(9455 / 12861)  [      ] group1/block1/conv1/bn/variance/EMA                                                                                                   
(9456 / 12861)  [      ] group1/block1/conv1/bn/variance/EMA/read                                                                                              
(9457 / 12861)  [      ] tower0/group1/block1/conv1/bn/Reshape_3/shape                                                                                         
(9458 / 12861)  [      ] tower0/group1/block1/conv1/bn/Reshape_3                                                                                               
(9459 / 12861)  [      ] tower0/group1/block1/conv1/bn/batchnorm/add/y                                                                                         
(9460 / 12861)  [      ] tower0/group1/block1/conv1/bn/batchnorm/add                                                                                           
(9461 / 12861)  [      ] tower0/group1/block1/conv1/bn/batchnorm/Rsqrt                                                                                         
(9462 / 12861)  [      ] group1/block1/conv1/bn/gamma                                                                                                          
(9463 / 12861)  [      ] group1/block1/conv1/bn/gamma/read                                                                                                     
(9464 / 12861)  [      ] tower0/group1/block1/conv1/bn/Reshape/shape                                                                                           
(9465 / 12861)  [      ] tower0/group1/block1/conv1/bn/Reshape                                                                                                 
(9466 / 12861)  [      ] tower0/group1/block1/conv1/bn/batchnorm/mul                                                                                           
(9467 / 12861)  [      ] tower0/group1/block1/conv1/bn/batchnorm/mul_1                                                                                         
(9468 / 12861)  [      ] tower0/gradients/tower0/group1/block1/conv1/bn/batchnorm/add_1_grad/Shape                                                             
(9469 / 12861)  [      ] tower0/gradients/tower0/group1/block1/conv1/bn/batchnorm/add_1_grad/BroadcastGradientArgs                                             
(9470 / 12861)  [      ] tower0/group1/block1/conv1/bn/batchnorm/mul_2                                                                                         
(9471 / 12861)  [      ] tower0/group1/block1/conv1/bn/batchnorm/sub                                                                                           
(9472 / 12861)  [      ] tower0/group1/block1/conv1/bn/batchnorm/add_1                                                                                         
(9473 / 12861)  [      ] tower0/group1/block1/conv1/bn/output
```
Some has `tower0/` in the front and some don't. Also all nodes begin with `apply_gradients` don't have tower prefix. For example:
```
     (669 / 12861)  [      ] apply_gradients/AccumGrad_1/counter                                                                                                    
     (670 / 12861)  [      ] apply_gradients/AccumGrad_1/counter/read                                                                                               
     (671 / 12861)  [      ] apply_gradients/AccumGradOptimizer_1/FloorMod                                                                                          
     (672 / 12861)  [      ] apply_gradients/AccumGradOptimizer_1/Equal                                                                                             
     (673 / 12861)  [      ] apply_gradients/AccumGradOptimizer_1/update_counter/value                                                                              
     (674 / 12861)  [      ] apply_gradients/AccumGradOptimizer_1/update_counter  
```
Does this mean that these nodes are not part of the tower function? How could some nodes of one layer be in the tower and some not?

Motivation: I get a lot of `inf` after several run with my own model, yet debugging it is really confusing as I can't understand the meaning of name of the nodes (due to my lack of knowledge of inner works of tensorflow). For example, many nodes end with "Momentum" in my graph do not have "tower" prefixes, yet in the FasterRCNN example, they do have them. This makes me nervous as my implementation may have messed up with the name_scope staffs. May be that is the reason for the `inf`s?
I would be really grateful if you could share some idea on the name_scope mechanism of tensorpack, or some insights on debugging.  

Version:
Tensorpack:0.9.4

Thank you very much.

**Comment by ppwwyyxx** [comment]
`group1/block1/conv1/bn/beta` does not have tower prefix because it corresponds to variables. This is by design in tower0, so they will have a clean name when saved as checkpoints.

Tower function is responsible for computing the gradients. Therefore `apply_gradients*` won't have the tower prefix because they are not part of the tower function.

**Comment by zhou-git** [comment]
Hi,

Is there any elaborate on the difference between layer_register and under_name_scope?

Also,
In the tutorial on tower function, the name of a variable is like 
> The name of any trainable variables created in the function must be like “variable_scope_name/custom/scopes/name”. Therefore, the name of any trainable variables must:
Not depend on name_scope’s name.
Not depend on some tensor’s name.
Not use the same variable_scope’s name twice.
Tensorpack layers create variables based on the name given to the layer (i.e., Conv2D('name', x)). So the name of the layer needs to follow the above rules as well.

What is the specific meaning of each component of the whole name? I think the meaning of "name" is clear, "variable_scope_name" refers to towers? In FasterRCNN, we get names like 
```
Tensor("tower0/rpn_losses/label_loss:0", shape=(), dtype=float32, device=/device:GPU:0)
```
Is this a violation of the naming convention?

**Comment by ppwwyyxx** [comment]
> What is the specific meaning of each component of the whole name? 

e.g., `tower0`, `rpn_losses`, `label_loss` in your example. I'll update the docs to make it more clear.

> Is this a violation of the naming convention?

No because this is a Tensor, not a trainable variable, so it can have whatever name.
If it is a trainable variable, this is a valid name iff `tf.get_variable_scope() == 'tower0'`.

**Comment by ppwwyyxx** [comment]
> Is there any elaborate on the difference between layer_register and under_name_scope?

As far as names are concerned, `layer_register` opens a variable scope and `under_name_scope` opens a name scope.

**Comment by zhou-git** [comment]
Thanks for the quick reply. 

It is mentioned in 
```
build_graph, train.py:88
```
in `compute_grad_from_inputs`.

**Comment by ppwwyyxx** [comment]
Thanks. I'm not sure I understand this question correctly:

> What is the specific meaning of each component of the whole name?

Do you now have an answer already? The components comes from scopes and their meaning come from users who create the scopes, except for variable scopes like "tower0" which are created by tensorpack and are supposed to be respected.

**Comment by zhou-git** [comment]
> No because this is a Tensor, not a trainable variable, so it can have whatever name.
If it is a trainable variable, this is a valid name iff tf.get_variable_scope() == 'tower0'.

> Do you now have an answer already? 

Thanks.
I guess it is cleared a bit. 
I think the question comes down to: what is the rules must be respected(I guess this is the correct word?) when constructing custom layers.
For variables: you are saying for a trainable variable, as long as the variable scope is correct, then it should be OK? So even I messed up the `custom/scopes/name` part, it does not affect the correctness of the gradients?
And for tensors(layers), any name will suffice?
